### ISSUE TITLE
Unable to see the list of Stubs in the Dynamic Stubs section

### ISSUE THREAD

**Issue by panbhatt** [i-anunay/mock-node#7]
Scenerio:
1. i add a number of STUBS and created routes. 
   ![image](https://cloud.githubusercontent.com/assets/1473470/15973082/90c329ca-2f0f-11e6-98e3-898c617e0ef3.png)
2. i tried to create the Dynamic stubs URL, but ones i entered the URI and go in manage dynamic stubs, the list of stubs doesn't appear. 
   ![image](https://cloud.githubusercontent.com/assets/1473470/15973118/b41e0fde-2f0f-11e6-89bb-15fbc91f8c46.png)

The Select list in the above is EMPTY. Is this is a expected behavior or there is a mismatch. I even restarted the server. I am running on WINDOWS 7.

**Comment by ianunay** [comment]
Hi,

Stubs and Dynamic stubs only exist in the context of the route.

Stubs created for /employee/ route will only be available for the /employee/ handlers and dynamic stub configuration of /employee/ route.

In the issue mentioned above notice that the route has changed: /employee/ and /demployee/
stubs created in /employee/ will not be available in /demployee/

you can create dynamic stubs for the same /employee/ route and use handlers to change how that route behaves, rather than creating two routes.

**Comment by panbhatt** [comment]
Thanks. That's little confusion though. I tried it and now I am getting an exception. 

![image](https://cloud.githubusercontent.com/assets/1473470/15974848/a9c0a1c4-2f18-11e6-92e1-fcc88174bcf0.png)

i tried http://localhost:8000/employee and now it is throwing an exception. 

TypeError: Cannot read property 'eval' of undefined
   at C:\Temp\mock-node\server.js:160:138
   at Object.async.whilst (C:\Temp\mock-node\node_modules\async\lib\async.js:792:13)
   at C:\Temp\mock-node\server.js:154:15
   at Layer.handle [as handle_request](C:Tempmock-nodenode_modulesexpresslibrouterlayer.js:95:5)
   at trim_prefix (C:\Temp\mock-node\node_modules\express\lib\router\index.js:312:13)
   at C:\Temp\mock-node\node_modules\express\lib\router\index.js:280:7
   at Function.process_params (C:\Temp\mock-node\node_modules\express\lib\router\index.js:330:12)
   at next (C:\Temp\mock-node\node_modules\express\lib\router\index.js:271:10)
   at C:\Temp\mock-node\server.js:85:16
   at Layer.handle [as handle_request](C:Tempmock-nodenode_modulesexpresslibrouterlayer.js:95:5)

Even the default stub is not returned in case of /employee

I m assuing different stub in case of /employee and /employee?gender=male

**Comment by ianunay** [comment]
Looks like the condition wasn't saved properly in the config for some reason, can you check the dynamic stub config and see that there are no empty strings in the conditions.

**Comment by panbhatt** [comment]
i think it is stored properly. I like the concept, but it will be awesome, it we can make it much simpler.. 

Here is the config. 
 "dynamicStubs": [
        {
          "name": "P1",
          "description": "",
          "defaultStub": "Product1.json",
          "conditions": [
            {
              "eval": "req.query.avail=='true'",
              "stub": "Product2.json"
            }
          ]
        },
        {
          "name": "P2",
          "description": "",
          "defaultStub": "Product1.json",
          "conditions": [
            {
              "eval": "req.query.avail=='false'",
              "stub": "Product1.json"
            }
          ]
        }
      ],

I prefer this approach. 
1. Create a global set of JSON's and name them
2. Configure the Routes to point to any one of them. 

This will help to simplify the approach as there is a global pool of JSON  documents.

**Comment by panbhatt** [comment]
Any update on this. I recently tried with a new fresh installation. This time i have query params like this req.query.age > 24 , then I am showing a stub otherwise default stub. but it keep on giving on ERROR. 

TypeError: Cannot read property 'eval' of undefined
   at D:\Project\Mine\MockNode\mock-node\server.js:160:138
   at Object.async.whilst (D:\Project\Mine\MockNode\mock-node\node_modules\async\lib\async.js:792:13)
   at D:\Project\Mine\MockNode\mock-node\server.js:154:15
   at Layer.handle [as handle_request](D:ProjectMineMockNodemock-nodenode_modulesexpresslibrouterlayer.js:95:5)
   at trim_prefix (D:\Project\Mine\MockNode\mock-node\node_modules\express\lib\router\index.js:312:13)
   at D:\Project\Mine\MockNode\mock-node\node_modules\express\lib\router\index.js:280:7
   at Function.process_params (D:\Project\Mine\MockNode\mock-node\node_modules\express\lib\router\index.js:330:12)
   at next (D:\Project\Mine\MockNode\mock-node\node_modules\express\lib\router\index.js:271:10)
   at D:\Project\Mine\MockNode\mock-node\server.js:85:16
   at Layer.handle [as handle_request](D:ProjectMineMockNodemock-nodenode_modulesexpresslibrouterlayer.js:95:5)

**Comment by ianunay** [comment]
Can you please paste the entire config.json file here? It looks like one of the condition's eval property is undefined for some reason.

**Comment by panbhatt** [comment]
{
  "port": 3000,
  "routes": [
    {
      "route": "/users/",
      "handle": "proxy",
      "proxy": "http://localhost:8090",
      "stub": "users.json",
      "dynamicStub": "user conditions",
      "stubs": [
        {
          "name": "admins.json",
          "description": "A list of admins, this returns an array."
        },
        {
          "name": "users.json",
          "description": "this is the general users"
        }
      ],
      "dynamicStubs": [
        {
          "name": "user conditions",
          "description": "this is an example conditional stubing",
          "defaultStub": "id.json",
          "conditions": [
            {
              "eval": "req.body.user == 'admin'",
              "stub": "admins.json"
            },
            {
              "eval": "req.query.name == 'anunay'",
              "stub": "users.json"
            }
          ]
        }
      ]
    },
    {
      "route": "/id/",
      "handle": "dynamicStub",
      "proxy": "http://localhost:8090",
      "stub": "id.json",
      "dynamicStub": "random conditions",
      "stubs": [
        {
          "name": "id.json",
          "description": "A list of ids"
        },
        {
          "name": "newStub.json",
          "description": "dummy stub for example"
        }
      ],
      "dynamicStubs": [
        {
          "name": "random conditions",
          "description": "this is conditional stubing for ids",
          "defaultStub": "id.json",
          "conditions": [
            {
              "eval": "req.body.id == 'id'",
              "stub": "id.json"
            },
            {
              "eval": "req.body.newStub == 'new'",
              "stub": "newStub.json"
            }
          ]
        }
      ]
    },
    {
      "route": "/index/",
      "handle": "stub",
      "proxy": "http://localhost:8090",
      "stub": "index.xml",
      "dynamicStub": "",
      "stubs": [
        {
          "name": "index.html",
          "description": "You can even stub html files"
        },
        {
          "name": "index.xml",
          "description": "This is an example xml file"
        }
      ],
      "dynamicStubs": []
    },
    {
      "route": "/employee/",
      "handle": "dynamicStub",
      "proxy": "http://localhost:3000",
      "stubs": [
        {
          "name": "EMPLOYEE1",
          "description": ""
        },
        {
          "name": "EMPLOYEEP",
          "description": ""
        }
      ],
      "dynamicStubs": [
        {
          "name": "Dynamic1",
          "defaultStub": "EMPLOYEE1",
          "conditions": [
            {
              "stub": "EMPLOYEEP",
              "eval": "req.query.age > 25"
            }
          ]
        }
      ],
      "dynamicStub": "Dynamic1"
    }
  ],
  "global": {
    "delay": 0,
    "headers": {
      "Access-Control-Allow-Origin": "*",
      "X-Powered-By": "mocknode"
    }
  }
}

**Comment by panbhatt** [comment]
i still can't understand this piece of code, Since i have only one condition

 function (callback) {
          count++;
          if (count == _stub.conditions.length - 1) continueLoop = false;

```
      var script = sandcastle.createScript('exports.main = function() {\n          try {\n            if (' + _stub.conditions[count].eval + ')\n              exit(\'' + _stub.conditions[count].stub + '\')\n            else\n              exit(false)\n          } catch(e) {\n            exit(false)\n          }\n        }');
```

and count is increasing to 1 , the first element does not exist in the array, so the error is coming. 

I AM ON MASTER Branch???

**Comment by ianunay** [comment]
see [mock-node/src/server.es6](https://github.com/ianunay/mock-node/blob/master/src/server.es6)

server.js is transpiled from server.es6 from src. by >> npm run build-server

yeah looks like the count increment is being done at the beginning, it should be probably done at the end of the loop and the if condition also has to be corrected.

**Comment by panbhatt** [comment]
did you make it work???

**Comment by panbhatt** [comment]
FYI
I am running node 5.8.0 and the sandcastle script is not running at all. It never goes in the exit function. I had to manually write the script.on('timeout') and i can see it is always going to timeout. I dint know how you had tested it ??

**Comment by ianunay** [comment]
It works on ubuntu , node v5.0.0 and you can also test on the heroku deployed version, it works there.

**Comment by panbhatt** [comment]
Okies. I will try to deploy it on local ubuntu. I was working on WINDOWS. May be the sandcastle doesn't have support for windows.

**Comment by ianunay** [comment]
I've just pushed a fix for the issue with the conditions loop,  now the proper stubs are returned. I've added the sandboxing functionality at the end and pushed without much testing , lesson learnt.

Sandcastle uses buffertools as a dependency and during npm install it runs a node-gyp, during the install did you happen to get any errors?

**Comment by panbhatt** [comment]
@ianunay this seems to be working now. I worked on UBUNTU and there are certain different steps that I took during this time. I noticed once we create a dynamic stub we need to go on the main window and update it. If we update it from INSIDE (where we enter the detail) it doesn't work, it has to be done from the main window. 

Some other things. 
1. Since we are storing JSON, currently the response header is octet stream , we can hard code this thing for the time being as JSON as it is the only widely use transfer mechanism now adays.
